src/goCheck.ts

```diff
@@ -142,20 +142,41 @@ export function check(filename: string, goConfig: vscode.WorkspaceConfiguration)
 		return testPromise;
 	};
 
-	if (!!goConfig['buildOnSave']) {
+	if (!!goConfig['buildOnSave'] && goConfig['buildOnSave'] !== 'off') {
 		let buildFlags = goConfig['buildFlags'] || [];
 		let buildTags = '"' + goConfig['buildTags'] + '"';
-		let tmppath = path.normalize(path.join(os.tmpdir(), 'go-code-check'));
+
 		let currentGoWorkspace = getCurrentGoWorkspaceFromGOPATH(cwd);
-		let importPath = currentGoWorkspace ? cwd.substr(currentGoWorkspace.length + 1) : '.';
-		let args = ['build', '-i', '-o', tmppath, '-tags', buildTags, ...buildFlags, importPath];
+		let tmppath = path.normalize(path.join(os.tmpdir(), 'go-code-check'));
+
+		let buildWorkDir = cwd;
+		let buildWorkspace = goConfig['buildOnSave'] === 'workspace';
+
+		let args: any[];
 		if (filename.match(/_test.go$/i)) {
-			args = ['test', '-i', '-copybinary', '-o', tmppath, '-c', '-tags', buildTags, ...buildFlags, importPath];
+			args = ['test', '-i'];
+
+			if (!buildWorkspace) {
+				args.push('-copybinary', '-o', tmppath, '-c');
+			}
+		} else {
+			args = ['build', '-i'];
+
+			if (!buildWorkspace) {
+				args.push('-o', tmppath);
+			}
+		}
+
+		args.push('-tags', buildTags, ...buildFlags);
+
+		if (buildWorkspace) {
+			buildWorkDir = vscode.workspace.rootPath;
+			args.push('./...');
```

Comment:
When the current file that is open is a test file, this will end up running `go test -i ./...` which will run the tests in the workspace instead of trying to compile them

For compiling test files, we need to use `go test -c` and `-c` cannot be used on multi-package.

@rakyll @campoy @mattetti Do you know of a way to compile test files under a directory?